### PR TITLE
s390x: Hard code the link register for ElfTlsGetOffset

### PR DESCRIPTION
**Pull request by bjorn3**
The ElfTlsGetOffset pseudo-instruction already hard codes the rest of the abi anyway and if another abi is ever needed, adding another pseudo-instruction or an enum field to the existing one would likely be necessary anyway. Also remove a couple of dead_code allows.

### CODE REVIEW

**Comment by cfallin** [comment]
cc @uweigand -- this looks reasonable to me overall, and there's no functional change to the emitted code or the regalloc metadata (fields were only ever used with one value) -- but maybe there's a reason it was generic?

**Comment by uweigand** [comment]
Well, as long as the ELF ABI is the only s390x ABI supported by wasmtime, I guess it doesn't make any difference, really.

That said, this seems a step in the wrong direction to me: there *are* other ABIs used by other OSes on the platforms, and they do use different registers than r14 as the link register, so I'm not sure why it's an improvement to hard-code that value?  Logically, it is an ABI property, that's why I placed that choice in ABI code rather than ISA code.

**Comment by bjorn3** [comment]
The ret instruction is also hard coding r15 (stack pointer?) and it has a debug assert that r14 is used as link register. What other OSes use a different link register?

**Comment by uweigand** [comment]
> The ret instruction is also hard coding r15 (stack pointer?) 

Ah, not really.  This instruction uses a condition mask field as "R1", not an actual register field, and the contents are 15 (all mask bits set) to indicate a "branch on any condition", i.e. an unconditional branch.  It would have been clearer to add a variant of the `enc_rr` helper instead of passing the condition mask as a "register".

> and it has a debug assert that r14 is used as link register.

Huh, I'm surprised to see that.  Looks like @elliottt added a bunch of asserts here: https://github.com/bytecodealliance/wasmtime/pull/5172 - I'm not sure if there's anything in that diff that makes those asserts necessary?

> What other OSes use a different link register?

Well, XPLINK (the current main calling convention on z/OS) uses %r4, as one major example.


In general, I'm pretty sure in case we ever wanted to support a significantly different calling convention like XPLINK that uses different registers for the link register, the stack register, and pretty much any ABI-defined register, we will find places where the current code isn't fully generalized.  I tried to keep things general where it seemed easily possible, but didn't attempt any full review ...   I just don't see why it's beneficial now to make the code *less* general.

**Comment by elliottt** [comment]
> Huh, I'm surprised to see that. Looks like @elliottt added a bunch of asserts here: #5172 - I'm not sure if there's anything in that diff that makes those asserts necessary?

I think you're right, we should probably remove them. They're added in other cases to check that fixed constraints are respected, but as there's no fixed constraints in the constraint generation code in mod.rs, I don't think that assertion applies.

**Comment by uweigand** [comment]
> > Huh, I'm surprised to see that. Looks like @elliottt added a bunch of asserts here: #5172 - I'm not sure if there's anything in that diff that makes those asserts necessary?
> 
> I think you're right, we should probably remove them. They're added in other cases to check that fixed constraints are respected, but as there's no fixed constraints in the constraint generation code in mod.rs, I don't think that assertion applies.

Thanks.  I've now submitted a PR: https://github.com/bytecodealliance/wasmtime/pull/9397

**Comment by bjorn3** [comment]
I removed the call instruction changes. I think it still makes sense to keep them for the ElfTlsGetOffset pseudo-instruction as that one hard codes a single abi anyway.

File: cranelift/codegen/src/isa/s390x/inst/regs.rs [hunk 1]
```diff
@@ -78,6 +78,10 @@ pub fn zero_reg() -> Reg {
     gpr(0)
 }
 
+pub fn link_reg() -> Reg {
```
Comment (uweigand) [inline cranelift/codegen/src/isa/s390x/inst/regs.rs:81]: I don't really like this.  This function, with this name in this place, gives the wrong impression that this is a ISA-defined register, and encourages additional uses in the future.

If we need a function like this, it should be an *ABI* function.

In any case, if you really want to hard-code the link register in `ElfTlsGetOffset`, then I'd suggest to just hard-code 14 directly, like this code also hard-codes register numbers 12 and 2.  (Again, I'd prefer to go the other way and not hard-code any of them, and instead get them from ABI code.  But given that this opcode is ABI-specific, I don't really have a strong opinion here.)

Comment (bjorn3) [reply cranelift/codegen/src/isa/s390x/inst/regs.rs]: Inlined in all callers.

Comment (uweigand) [reply cranelift/codegen/src/isa/s390x/inst/regs.rs]: LGTM now, thanks!

**Review by cfallin — APPROVED** [review]

**Pull request merged**